node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+                            ) => {
+                                let msg = BitfieldGossipMessage {
+                                    relay_parent: hash,
+                                    signed_availability,
+                                };
+                                // @todo do we also make sure to not send it twice if we are the source
+                                // @todo this also apply to ourself?
+                                distribute(ctx.clone(), &mut tracker, msg).await?;
+                            }
+                            BitfieldDistributionMessage::NetworkBridgeUpdate(event) => {
+                                handle_network_msg(ctx.clone(), &mut tracker, event).await;
+                            }
+                        }
+                    }
+                    FromOverseer::Signal(OverseerSignal::StartWork(relay_parent)) => {
+                        let (validator_set, signing_context) =
+                            query_basics(ctx.clone(), relay_parent).await?;
+
+                        let _ = tracker.per_relay_parent.insert(
+                            relay_parent,
+                            PerRelayParentData {
+                                signing_context,
+                                validator_set: validator_set,
+                                ..Default::default()
+                            },
+                        );
+                        // futurama.insert(relay_parent.clone(), Box::pin(future.map(|_| ())));
+                        // active_jobs.insert(relay_parent.clone(), abort_handle);
+                    }
+                    FromOverseer::Signal(OverseerSignal::StopWork(relay_parent)) => {
+                        let _ = tracker.per_relay_parent.remove(&relay_parent);
+                    }
+                    FromOverseer::Signal(OverseerSignal::Conclude) => {
+                        tracker.per_relay_parent.clear();
+                        return Ok(());
+                    }
+                }
+            }
+            // active_jobs
+            //     .retain(|_, future| future.poll().is_pending());
+        }
+    }
+}
+
+/// modify the reputiation, good or bad
+async fn modify_reputiation<Context>(
+    mut ctx: Context,
+    peerid: PeerId,
+    rep: ReputationChange,
+) -> SubsystemResult<()>
+where
+    Context: SubsystemContext<Message = BitfieldDistributionMessage> + Clone,
+{
+    ctx.send_message(AllMessages::NetworkBridge(
+        NetworkBridgeMessage::ReportPeer(peerid, rep),
+    ))
+    .await
+}
+
+/// Distribute a checked message, either originated by us or gossiped on from other peers.
+async fn distribute<Context>(
+    mut ctx: Context,
+    tracker: &mut Tracker,
+    message: BitfieldGossipMessage,
+) -> SubsystemResult<()>
+where
+    Context: SubsystemContext<Message = BitfieldDistributionMessage> + Clone,
+{
+
+    let BitfieldGossipMessage {
+        relay_parent,
+        signed_availability,
+    } = message;
```

Comment:
If we just want to unpack a struct, we may as well do it in the function signature.

```suggestion
    BitfieldGossipMessage { relay_parent, signed_availability }: BitfieldGossipMessage,
) -> SubsystemResult<()>
where
    Context: SubsystemContext<Message = BitfieldDistributionMessage> + Clone,
{
```

Reply:
I wasn't aware this was possible!

Reply:
Rust's pattern-matching capabilities are kind of amazing.